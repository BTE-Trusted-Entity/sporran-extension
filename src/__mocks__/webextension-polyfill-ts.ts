// src/__mocks__/webextension-polyfill-ts
// Update this file to include any mocks for the `webextension-polyfill-ts` package
// This is used to mock these values for Storybook so you can develop your components
// outside the Web Extension environment provided by a compatible browser

import { setupGetMessageShim } from 'chrome-extension-i18n-shim';

import messagesEN from '../static/_locales/en/messages.json';

export const browser = {
  tabs: {
    executeScript(
      currentTabId: number /* eslint-disable-line @typescript-eslint/no-unused-vars */,
      details: unknown /* eslint-disable-line @typescript-eslint/no-unused-vars */,
    ): Promise<{ done: boolean }> {
      return Promise.resolve({ done: true });
    },
  },
  i18n: {
    getMessage: setupGetMessageShim(messagesEN),
    getUILanguage(): string {
      return document.documentElement.lang || 'en';
    },
  },
  storage: {
    local: {
      async get(): Promise<unknown> {
        return {
          authorizedDApps: {
            'example.com': false,
            'examplesuperlong.dfghrtydfghklyuihjwertyxcvbn3456rty56782345678yudfgghtyuwertyudfghsdfgh-domain.org':
              true,
            'example.org': true,
          },
        };
      },
      async set(): Promise<void> {
        // dummy
      },
    },
  },
  runtime: {
    async sendMessage(): Promise<void> {
      // dummy
    },
    onMessage: {
      addListener(): void {
        // dummy
      },
      removeListener(): void {
        // dummy
      },
    },
  },
};

export interface Tabs {
  Tab: {
    id: number;
  };
}
