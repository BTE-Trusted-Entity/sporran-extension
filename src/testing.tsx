export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
