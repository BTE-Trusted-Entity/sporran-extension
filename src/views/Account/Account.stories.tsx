import { Meta } from '@storybook/react';

import { Account } from './Account';

export default {
  title: 'Views/Account',
  component: Account,
} as Meta;

export function Template(): JSX.Element {
  return (
    <Account
      account={{
        name: 'My Sporran Account',
        address: '4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire',
      }}
    />
  );
}
