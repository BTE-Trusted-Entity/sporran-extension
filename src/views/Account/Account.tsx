import { useCallback, useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { browser } from 'webextension-polyfill-ts';

import { Balance } from '../../components/Balance/Balance';
import {
  saveAccount,
  setCurrentAccount,
} from '../../utilities/accounts/accounts';
import { generatePath, paths } from '../paths';

interface Props {
  account: {
    address: string;
    name: string;
    index: number;
  };
}

export function Account({ account }: Props): JSX.Element {
  const t = browser.i18n.getMessage;

  useEffect(() => {
    setCurrentAccount(account.address);
  }, [account]);

  const [editing, setEditing] = useState(false);

  const handleEditClick = useCallback(() => {
    setEditing(true);
  }, []);

  const handleCancelClick = useCallback(() => {
    setEditing(false);
  }, []);

  const handleSubmit = useCallback(
    async (event) => {
      event.preventDefault();

      const name = event.target.elements.name.value;
      await saveAccount({
        ...account,
        name,
      });

      setEditing(false);
    },
    [account],
  );

  return (
    <section>
      <h2>{account.name}</h2>

      {!editing && (
        <button type="button" onClick={handleEditClick}>
          {t('view_Account_rename')}
        </button>
      )}

      {editing && (
        <form onSubmit={handleSubmit}>
          <label>
            {t('view_Account_name')}
            <input name="name" required />
          </label>
          <button type="submit">{t('common_action_save')}</button>

          <button type="button" onClick={handleCancelClick}>
            {t('common_action_cancel')}
          </button>
        </form>
      )}

      <p>
        <Balance address={account.address} />
      </p>

      <p>
        <Link to={generatePath(paths.account.send, account)}>
          {t('view_Account_send')}
        </Link>
      </p>
      <p>
        <Link to={generatePath(paths.account.receive, account)}>
          {t('view_Account_receive')}
        </Link>
      </p>
    </section>
  );
}
