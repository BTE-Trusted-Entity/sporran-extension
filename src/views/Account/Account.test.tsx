import userEvent from '@testing-library/user-event';
import { browser } from 'webextension-polyfill-ts';

import { render, screen } from '../../testing';
import {
  saveAccount,
  setCurrentAccount,
} from '../../utilities/accounts/accounts';
import {
  BalanceChangeResponse,
  MessageType,
} from '../../connection/MessageType';

import { Account } from './Account';

jest.mock('../../utilities/accounts/accounts');
jest.spyOn(browser.runtime, 'sendMessage');
jest
  .spyOn(browser.runtime.onMessage, 'addListener')
  .mockImplementation(async (callback) => {
    const response = {
      type: MessageType.balanceChangeResponse,
      data: { balance: '04625103a72000' },
    } as BalanceChangeResponse;
    callback(response, {});
  });

const account = {
  name: 'My Sporran Account',
  address: '4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire',
  index: 1,
};

describe('Account', () => {
  it('should render', async () => {
    const { container } = render(<Account account={account} />);
    expect(container).toMatchSnapshot();
  });

  it('should update the current account', async () => {
    render(<Account account={account} />);
    expect(setCurrentAccount).toHaveBeenCalledWith(account.address);
  });

  it('should enable editing the account name', async () => {
    render(<Account account={account} />);

    userEvent.click(await screen.findByRole('button', { name: 'Rename' }));
    userEvent.type(await screen.findByLabelText('Account name:'), 'Foo');
    userEvent.click(await screen.findByRole('button', { name: 'Save' }));

    expect(
      await screen.findByRole('button', { name: 'Rename' }),
    ).toBeInTheDocument();

    expect(saveAccount).toHaveBeenCalledWith({
      name: 'Foo',
      address: account.address,
      index: 1,
    });
  });
});
