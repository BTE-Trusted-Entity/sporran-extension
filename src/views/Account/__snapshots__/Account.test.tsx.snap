// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account should render 1`] = `
<div>
  <section>
    <h2>
      My Sporran Account
    </h2>
    <button
      type="button"
    >
      Rename
    </button>
    <p>
      <span>
        Balance: 
        1.2340
         
        K
      </span>
    </p>
    <p>
      <a
        href="/account/4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire/send"
      >
        Send
      </a>
    </p>
    <p>
      <a
        href="/account/4tJbxxKqYRv3gDvY66BKyKzZheHEH8a27VBiMfeGX2iQrire/receive"
      >
        Receive
      </a>
    </p>
  </section>
</div>
`;
